src/thor/trace_route_action.cc

```diff
@@ -394,92 +395,105 @@ thor_worker_t::map_match(Api& request) {
+                     std::vector<meili::MatchResult>::const_iterator,
+                     std::vector<meili::MatchResult>::const_iterator>;
+      std::vector<edge_group_t> edge_groups{
+          edge_group_t{path_edges.cbegin(), path_edges.cbegin(), {}, {}}};
       auto discontinuity_edges_iter = m_temp_disconnected_edges.begin();
-      for (int i = 1, n = static_cast<int>(m_temp_path_edges.size()); i < n; ++i) {
-        const auto& front_edge_info = m_temp_path_edges[i - 1];
-        const auto& back_edge_info = m_temp_path_edges[i];
+      for (auto path_edge_itr = std::next(path_edges.cbegin()); path_edge_itr != path_edges.cend();
+           ++path_edge_itr) {
+        auto prev_path_edge_itr = std::prev(path_edge_itr);
         // make a new group when discontinuity occurs
         if (discontinuity_edges_iter != m_temp_disconnected_edges.end() &&
-            discontinuity_edges_iter->first == front_edge_info.edgeid &&
-            discontinuity_edges_iter->second == back_edge_info.edgeid) {
-          m_temp_disjoint_edge_groups.emplace_back();
+            discontinuity_edges_iter->first == prev_path_edge_itr->edgeid &&
+            discontinuity_edges_iter->second == path_edge_itr->edgeid) {
+          // start a new one
+          edge_groups.emplace_back(edge_group_t{path_edge_itr, path_edge_itr, {}, {}});
           ++discontinuity_edges_iter;
         }
-        m_temp_disjoint_edge_groups.back().first.emplace_back(back_edge_info);
+        // remember where the last one ended
+        std::get<1>(edge_groups.back()) = path_edge_itr;
       }
 
-      // here we mark edges in the path that get "upgraded" to break points
-      // because they occur just before a discontinuity
-      auto discontinuity_point_iter = match_results.cbegin();
-      for (auto& disjoint_edge_group : m_temp_disjoint_edge_groups) {
-        const auto& last_edge_in_group = disjoint_edge_group.first.back();
-        // in each edge group, we only care about the last edge where the discontinuity happens
-        while (discontinuity_point_iter->edgeid != last_edge_in_group.edgeid) {
-          ++discontinuity_point_iter;
-        }
-        // for the last edge, the last match result on this edge is where discontinuity happens
-        while (discontinuity_point_iter->edgeid == last_edge_in_group.edgeid) {
-          ++discontinuity_point_iter;
+      // here we mark the original shape points that just after and just before the
+      // discontinuities so that in the path they get "upgraded" to break points
+      auto match_result_itr = match_results.cbegin();
+      for (auto& edge_group : edge_groups) {
+        // we know the discontinuity happens on this edge so we just need to
+        // find the first match result on this edge
+        auto first_edge_id = std::get<0>(edge_group)->edgeid;
+        while (match_result_itr != match_results.cend() &&
```

Comment:
i added bounds checking to all of these while loops, it might be logically impossible that we are dereferencing the past the end iterator but id rather be safe than sorry